### PR TITLE
feat: added opentelemetry instrumentation for vertexai

### PR DESCRIPTION
**Pull request by 5war00p**
closes https://github.com/traceloop/openllmetry/issues/242
/claim #249

### CODE REVIEW

**Comment by 5war00p** [comment]
@nirga VertexAI has Text, Tabular, Image, Video and the Custom of creating datasets, training, model evaluation, getting predictions. So do we need to have traces for all the methods? or just getting predictions only?

**Comment by nirga** [comment]
@5war00p thanks! Supporting predictions API is sufficient for now. But make sure to test it thoroughly, including reporting of prompts and completions - similar to other APIs we have (OpenAI, etc.)

**Comment by 5war00p** [comment]
Understood. Btw from `language_models` every model has `predict` function (with interfaces related to code generation and completion) there is no completion function but code models has something called code completion.

**Comment by nirga** [comment]
I think the completion is what Google calls "generate_content". It's important to make sure that Gemini is supported and working is this is the one most wildly used.

**Comment by 5war00p** [comment]
Okies, I think this is what we are looking for: https://cloud.google.com/python/docs/reference/aiplatform/latest/vertexai.language_models.TextGenerationModel

It has `predict, predict_async, predict_streaming, predict_streaming_async` which we need right now.

**Comment by nirga** [comment]
@5war00p and this one - https://cloud.google.com/vertex-ai/docs/generative-ai/start/quickstarts/quickstart-multimodal

**Comment by 5war00p** [comment]
@nirga  What __init__.py for vertexai? I see they have this: https://cloud.google.com/python/docs/reference/spanner/3.18.0/opentelemetry-tracing
same as OpenAI have one.

**Comment by 5war00p** [comment]
image

image

**Comment by nirga** [comment]
@5war00p not sure I understand the question?

**Comment by 5war00p** [comment]
I'm saying there is an open-telemetry package for vertexai (google-cloud-aiplatform), so can we use that directly for span attributes similar to OpenAI (which we are doing in the repo) or we go with the custom way?

**Comment by 5war00p** [comment]
@nirga  Did you get what I was saying?

**Comment by nirga** [comment]
@5war00p no, let's move to slack maybe? https://traceloop.com/slack

**Comment by 5war00p** [comment]
The gemini output:

image

image

**Comment by 5war00p** [comment]
Streaming one here:
image

File: packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py [hunk 1]
```diff
@@ -1 +1,347 @@
...
+        "object": "GenerativeModel",
+        "method": "generate_content",
+        "span_name": "vertexai.generate_content",
+    },
+    {
+        "package": "vertexai.preview.generative_models",
+        "object": "GenerativeModel",
+        "method": "__init__",
+        "span_name": "vertexai.__init__",
+    },
```
Comment (galkleinman) [inline packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py:41]: no need to wrap?

File: packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py [hunk 2]
```diff
@@ -1 +1,347 @@
...
+        "object": "TextGenerationModel",
+        "method": "predict",
+        "span_name": "vertexai.predict",
+    },
+    {
+        "package": "vertexai.language_models",
+        "object": "TextGenerationModel",
+        "method": "from_pretrained",
+        "span_name": "vertexai.from_pretrained",
+    },
```
Comment (galkleinman) [inline packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py:53]: no need to wrap?

**Review by galkleinman — COMMENTED** [review]
Please fix the points mentioned here and the issue you've raised in the Community Slack.

I'll re-review afterward. these are only initial thoughts.

File: packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py [hunk 3]
```diff
@@ -1 +1,347 @@
...
+)
+
+from opentelemetry.semconv.ai import SpanAttributes, LLMRequestTypeValues
+from opentelemetry.instrumentation.vertexai.version import __version__
+
+from vertexai.generative_models._generative_models import Part
+
+logger = logging.getLogger(__name__)
+
+_instruments = ("google-cloud-aiplatform >= 1.38.1",)
```
[the remark this answers is not in the document]
Comment (5war00p) [reply packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py:27]: no reasons exactly, just followed the current version.

File: packages/traceloop-sdk/tests/test_vertexai_instrumentation.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                "what is shown in this image?",
+            ]
+        )
+        return response.text
+
+    generate_text()
+
+    spans = exporter.get_finished_spans()
+    assert [span.name for span in spans] == [
+        'POST',
```
[the remark this answers is not in the document]
Comment (5war00p) [reply packages/traceloop-sdk/tests/test_vertexai_instrumentation.py:38]: Added it to Exclude Urls, and removing POST.

File: packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py [hunk 5]
```diff
@@ -1 +1,347 @@
...
+                SpanAttributes.LLM_VENDOR: "VertexAI",
+                SpanAttributes.LLM_REQUEST_TYPE: LLMRequestTypeValues.COMPLETION.value,
+            }
+        ) as span:
+    
+        global llm_model
+
+
+        if (to_wrap.get('method') == 'from_pretrained' or to_wrap.get('method') == "__init__") and args is not None and len(args) > 0:
+            llm_model = args[0]
```
[the remark this answers is not in the document]
Comment (5war00p) [reply packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py:265]: We dont get model name in the response, I need to do it like this as only from_pretrained and __init__ has model names is that okey?

File: packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py [hunk 6]
```diff
@@ -1 +1,347 @@
...
+            prompt,
+        )
+
+    for key in kwargs:
+        if key in input_attribute_map:        
+            _set_span_attribute(
+                span,
+                input_attribute_map.get(key),
+                kwargs.get(key),
+            )
```
[the remark this answers is not in the document]
Comment (5war00p) [reply packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py:138]: Makes sense as there will be only 10 keys (or may be less)

File: packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py [hunk 7]
```diff
@@ -1 +1,347 @@
...
+                    # Gemini
+                    elif isinstance(subarg, Part):
+                        if hasattr(subarg, 'text'):
+                            prompt = f"{prompt}text: {subarg.text.file_uri}\n"
+                        elif hasattr(subarg, 'mime_type'):
+                            prompt = f"{prompt}mime_type: {subarg.mime_type}\n"
+                        elif hasattr(subarg, 'inline_data'):
+                            prompt = f"{prompt}data: {subarg.inline_data.data}\nmime_type: {subarg.inline_data.mime_type}\n"
+                        elif hasattr(subarg, 'file_data'):
+                            prompt = f"{prompt}file_uri: {subarg.file_data.file_uri}\nmime_type: {subarg.file_data.mime_type}\n"
```
[the remark this answers is not in the document]
Comment (5war00p) [reply packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py:122]: Yes, so vertexai package has something called `Part` and gemini model can take any sort of url (text, video, image..) with the mimetype. So we are checking that type and adding specific args in the prompt. May be we can directly add the object to the prompt (but this is more readable when user sees logs

File: packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (5war00p) [reply packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py:27]: What to use over here? I followed other instrumenters and they are doing the same

File: packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (galkleinman) [reply packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py]: you receive an argument using the wrapped methods of the current `instance` is it possible to extract the `model_name` out of that?

**Comment by 5war00p** [comment]
@nirga @galkleinman  Async Streaming Prediction working fine now. Also resolved review comments.

File: packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (5war00p) [reply packages/opentelemetry-instrumentation-vertexai/opentelemetry/instrumentation/vertexai/__init__.py]: For the `vertexai` we should always use `.from_trained` for PaLM2 and `GenerativeModel` for Gemini to pass model name, but you said to remove those two from the wrapped_methods list. 
https://github.com/traceloop/openllmetry/pull/279#discussion_r1435796808
https://github.com/traceloop/openllmetry/pull/279#discussion_r1435796845

**Comment by nirga** [comment]
Please rebase with main, there are a lot of unrelated changes in this PR.

**Comment by 5war00p** [comment]
Okey merged with main branch changes

**Comment by 5war00p** [comment]
Recreated this PR here: https://github.com/traceloop/openllmetry/pull/289

**Pull request closed without merging**